selfdrive/modeld/modeld.py

```diff
@@ -0,0 +1,307 @@
+
+  def run(self, buf: VisionBuf, wbuf: VisionBuf, transform: np.ndarray, transform_wide: np.ndarray,
+                inputs: Dict[str, np.ndarray], prepare_only: bool) -> Optional[np.ndarray]:
+    # Model decides when action is completed, so desire input is just a pulse triggered on rising edge
+    inputs['desire_pulse'][0] = 0
+    self.inputs['desire_pulse'][:-DESIRE_LEN] = self.inputs['desire_pulse'][DESIRE_LEN:]
+    self.inputs['desire_pulse'][-DESIRE_LEN:] = np.where(inputs['desire_pulse'] - self.prev_desire > .99, inputs['desire_pulse'], 0)
+    self.prev_desire[:] = inputs['desire_pulse']
+    logging.info("Desire enqueued")
+
+    self.inputs['traffic_convention'][:] = inputs['traffic_convention']
+    self.inputs['nav_features'][:] = inputs['nav_features']
+    self.inputs['nav_instructions'][:] = inputs['nav_instructions']
+    # self.inputs['driving_style'][:] = inputs['driving_style']
+
+    # if getCLBuffer is not None, frame will be None
+    frame = self.frame.prepare(buf, transform.astype(np.float32).flatten(), self.model.getCLBuffer("input_imgs"))
+    self.model.setInputBuffer("input_imgs", frame)
+    logging.info("Image added")
+
+    if wbuf is not None:
+      wide_frame = self.wide_frame.prepare(wbuf, transform_wide.astype(np.float32).flatten(), self.model.getCLBuffer("big_input_imgs"))
+      self.model.setInputBuffer("big_input_imgs", wide_frame)
+      logging.info("Extra image added")
+
+    if prepare_only:
+      return None
+
+    self.model.execute()
+    logging.info("Execution finished")
+
+    self.inputs['feature_buffer'][:-FEATURE_LEN] = self.inputs['feature_buffer'][FEATURE_LEN:]
+    self.inputs['feature_buffer'][-FEATURE_LEN:] = self.output[OUTPUT_SIZE:OUTPUT_SIZE+FEATURE_LEN]
+    logging.info("Features enqueued")
+
+    return self.output
+
+
+def main():
+  setproctitle("selfdrive.modeld.modeld")
+  if not PC:
+    set_realtime_priority(54)
+    set_core_affinity([7])
+
+  cl_context = CLContext()
+  model = ModelState(cl_context)
+  logging.warning("models loaded, modeld starting")
+
+  # visionipc clients
+  while True:
+    available_streams = VisionIpcClient.available_streams("camerad", block=False)
+    if available_streams:
+      use_extra_client = VisionStreamType.VISION_STREAM_WIDE_ROAD in available_streams and VisionStreamType.VISION_STREAM_ROAD in available_streams
+      main_wide_camera = VisionStreamType.VISION_STREAM_ROAD not in available_streams
+      break
+    time.sleep(.1)
+
+  vipc_client_main_stream = VisionStreamType.VISION_STREAM_WIDE_ROAD if main_wide_camera else VisionStreamType.VISION_STREAM_ROAD
+  vipc_client_main = VisionIpcClient("camerad", vipc_client_main_stream, True, cl_context)
+  vipc_client_extra = VisionIpcClient("camerad", VisionStreamType.VISION_STREAM_WIDE_ROAD, False, cl_context)
+  logging.warning(f"vision stream set up, main_wide_camera: {main_wide_camera}, use_extra_client: {use_extra_client}")
+
+  # TODO: Is it safe to use blocking=True here?
```

Comment:
don't think so because of msgq, but don't remember if this hits msgq yet. should be fine if you test it

Reply:
I think it doesn't exit when you block. @mitchellgoffpc Can you check it out and make sure the keyboardinterrupt works if that's the issue in a separate PR? That's a good fix anyway.

Reply:
Confirmed KeyboardInterrupt doesn't work with block=True, which makes sense since the blocking is handled in visionipc_client.cc. I think the only real way around this is to do the blocking in visionipc_pyx, but it's a bit ugly to have that logic duplicated...